pkg/command/root.go

```diff
@@ -320,6 +326,67 @@ func setupTargetPlugins() error {
 	return nil
 }
 
+// updateConfigWithTanzuCSPIssuer updates the "tanzu" and "mission-control" CLI contexts issuers with TCSP if the
+// issuer is VCSP Issuer, and invalidate the refresh token and token expiry time if these contexts token is
+// of the type id-token, so that CLI would re-trigger the interactive login with updated issuer.
+// This is done for only once.
+func updateConfigWithTanzuCSPIssuer(centralConfigIssuerUpdateFlagGetter func() bool,
+	cliContextUpdateStatusGetter func(string, interface{}) error) {
+
+	issuerUpdateFlag := centralConfigIssuerUpdateFlagGetter()
+	if !issuerUpdateFlag {
+		return
+	}
+
+	cliContextsCSPIssuerUpdated := false
+	_ = cliContextUpdateStatusGetter(isCLIContextsUpdatedToTCSPIssuers, &cliContextsCSPIssuerUpdated)
+	if cliContextsCSPIssuerUpdated {
+		return
+	}
+	cfg, err := config.GetClientConfig()
+	if err != nil {
+		return
+	}
+	if cfg == nil || len(cfg.KnownContexts) == 0 {
+		return
+	}
+	updateSuccess := true
+	for idx := range cfg.KnownContexts {
+		ctx := cfg.KnownContexts[idx]
+		if eligible := isEligibleForTCSPIssuerUpdate(ctx); !eligible {
+			continue
+		}
+		if ctx.GlobalOpts.Auth.Issuer == csp.StgIssuer {
+			ctx.GlobalOpts.Auth.Issuer = csp.StgIssuerTCSP
+		} else if ctx.GlobalOpts.Auth.Issuer == csp.ProdIssuer {
+			ctx.GlobalOpts.Auth.Issuer = csp.ProdIssuerTCSP
+		}
+		// invalidate only for interactive login token(id_token) and not for API Token type (API Tokens are carried over to TCSP)
+		if ctx.GlobalOpts.Auth.Type == csp.IDTokenType {
+			ctx.GlobalOpts.Auth.Expiration = time.Now().Local().Add(-10 * time.Second)
+			ctx.GlobalOpts.Auth.RefreshToken = "Invalid"
+		}
+		if err := config.SetContext(ctx, false); err != nil {
+			updateSuccess = false
+		}
+	}
+	// if all the contexts are updated successfully, update the flag in the data store
+	if updateSuccess {
+		_ = datastore.SetDataStoreValue(isCLIContextsUpdatedToTCSPIssuers, &updateSuccess)
```

Comment:
Should we show a log message to the user that all the contexts have been migrated to TCSP and tokens have been invalidated? That way, users might not be surprised when the interactive login happens again and it points to a new location.

Reply:
Good point. Adding the log is useful to inform user about the changes. I updated the PR with the new log message as shown below. Thanks
```
The CLI contexts have been updated to use the Tanzu CSP issuer. Any existing tokens obtained through interactive login are now invalid and CLI will automatically obtain a new token through interactive login using the new Tanzu CSP issuer
```